File: backend/app/routes/reviews.py
from flask import Blueprint, jsonify, request
from app import db
from app.models.review import Review

reviews_bp = Blueprint('reviews', __name__, url_prefix='/api/reviews')

@reviews_bp.route('/', methods=['GET'])
def get_reviews():
    """Récupérer tous les avis"""
    reviews = Review.query.all()
    return jsonify([r.to_dict() for r in reviews]), 200

@reviews_bp.route('/product/<int:product_id>', methods=['GET'])
def get_product_reviews(product_id):
    """Récupérer les avis d'un produit"""
    reviews = Review.query.filter_by(product_id=product_id).all()
    return jsonify([r.to_dict() for r in reviews]), 200

@reviews_bp.route('/', methods=['POST'])
def create_review():
    """Créer un avis"""
    data = request.get_json()
    
    if not all(k in data for k in ['product_id', 'user_id', 'rating']):
        return jsonify({'error': 'Champs requis: product_id, user_id, rating'}), 400
    
    if data['rating'] < 1 or data['rating'] > 5:
        return jsonify({'error': 'Note doit être entre 1 et 5'}), 400
    
    review = Review(
        product_id=data['product_id'],
        user_id=data['user_id'],
        rating=data['rating'],
        comment=data.get('comment')
    )
    
    db.session.add(review)
    db.session.commit()
    
    return jsonify(review.to_dict()), 201

@reviews_bp.route('/<int:review_id>', methods=['DELETE'])
def delete_review(review_id):
    """Supprimer un avis (admin)"""
    review = Review.query.get(review_id)
    
    if not review:
        return jsonify({'error': 'Avis non trouvé'}), 404
    
    db.session.delete(review)
    db.session.commit()
    
    return jsonify({'message': 'Avis supprimé'}), 200

File: backend/routes/reviews.js
const router = require('express').Router();
const db = require('../db');

const reviewQuery = `
  SELECT r.*, COALESCE(u.email, 'Anonyme') as user_email, p.name as product_name
  FROM reviews r
  LEFT JOIN users u ON r.user_id = u.id
  JOIN products p ON r.product_id = p.id
`;

router.get('/', async (req, res) => {
  try {
    const { rows } = await db.query(reviewQuery);
    res.json(rows);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/product/:product_id', async (req, res) => {
  try {
    const { rows } = await db.query(reviewQuery + ' WHERE r.product_id=$1', [req.params.product_id]);
    res.json(rows);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/', async (req, res) => {
  const { product_id, user_id, rating, comment, user_name } = req.body;
  if (!product_id || !rating)
    return res.status(400).json({ error: 'Champs requis: product_id, rating' });
  if (rating < 1 || rating > 5)
    return res.status(400).json({ error: 'Note doit être entre 1 et 5' });
  try {
    // Si pas de user_id, on insère avec NULL
    const { rows: [r] } = await db.query(
      'INSERT INTO reviews (product_id, user_id, rating, comment) VALUES ($1,$2,$3,$4) RETURNING id',
      [product_id, user_id || null, rating, comment]
    );
    const { rows: [review] } = await db.query(
      `SELECT r.*, 
        COALESCE(u.email, $2) as user_email,
        p.name as product_name
       FROM reviews r
       LEFT JOIN users u ON r.user_id = u.id
       JOIN products p ON r.product_id = p.id
       WHERE r.id=$1`,
      [r.id, user_name || 'Anonyme']
    );
    res.status(201).json(review);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.delete('/:id', async (req, res) => {
  try {
    const { rows } = await db.query('DELETE FROM reviews WHERE id=$1 RETURNING id', [req.params.id]);
    if (!rows[0]) return res.status(404).json({ error: 'Avis non trouvé' });
    res.json({ message: 'Avis supprimé' });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
